File: data_structures/heaps/Solutions.java
import java.util.*;

/** Heaps & Priority Queues — Min-Heap Implementation. */
public class Solutions {

    /** Binary min-heap backed by a flat array. O(log n) insert/extract. */
    public static class MinHeap {
        private final List<Integer> heap = new ArrayList<>();

        public void push(int val) {
            heap.add(val);
            siftUp(heap.size() - 1);
        }

        public int pop() {
            if (heap.isEmpty()) throw new NoSuchElementException("pop from empty heap");
            swap(0, heap.size() - 1);
            int val = heap.remove(heap.size() - 1);
            if (!heap.isEmpty()) siftDown(0);
            return val;
        }

        public int peek() { return heap.get(0); }
        public int size() { return heap.size(); }

        private void siftUp(int idx) {
            while (idx > 0) {
                int parent = (idx - 1) / 2;
                if (heap.get(idx) >= heap.get(parent)) break;
                swap(idx, parent);
                idx = parent;
            }
        }

        private void siftDown(int idx) {
            int n = heap.size();
            while (true) {
                int smallest = idx;
                int left = 2 * idx + 1, right = 2 * idx + 2;
                if (left < n && heap.get(left) < heap.get(smallest)) smallest = left;
                if (right < n && heap.get(right) < heap.get(smallest)) smallest = right;
                if (smallest == idx) break;
                swap(idx, smallest);
                idx = smallest;
            }
        }

        private void swap(int i, int j) {
            int tmp = heap.get(i);
            heap.set(i, heap.get(j));
            heap.set(j, tmp);
        }
    }

    /** Find K-th largest element using a min-heap of size k. O(n log k). */
    public static int findKthLargest(int[] nums, int k) {
        PriorityQueue<Integer> heap = new PriorityQueue<>();
        for (int num : nums) {
            heap.offer(num);
            if (heap.size() > k) heap.poll();
        }
        return heap.peek();
    }

    /** Merge K sorted lists using a priority queue. O(N log k) time. */
    public static List<Integer> mergeKSortedLists(List<List<Integer>> lists) {
        PriorityQueue<int[]> heap = new PriorityQueue<>((a, b) -> a[0] - b[0]);
        for (int i = 0; i < lists.size(); i++) {
            if (!lists.get(i).isEmpty()) {
                heap.offer(new int[]{lists.get(i).get(0), i, 0});
            }
        }
        List<Integer> result = new ArrayList<>();
        while (!heap.isEmpty()) {
            int[] curr = heap.poll();
            result.add(curr[0]);
            int listIdx = curr[1], elemIdx = curr[2];
            if (elemIdx + 1 < lists.get(listIdx).size()) {
                heap.offer(new int[]{lists.get(listIdx).get(elemIdx + 1), listIdx, elemIdx + 1});
            }
        }
        return result;
    }
}

File: data_structures/heaps/solutions.py
"""Heaps & Priority Queues — Min-Heap Implementation."""

from typing import List


class MinHeap:
    """Binary min-heap backed by a flat array. O(log n) insert/extract."""

    def __init__(self) -> None:
        self._heap: List[int] = []

    def push(self, val: int) -> None:
        self._heap.append(val)
        self._sift_up(len(self._heap) - 1)

    def pop(self) -> int:
        if not self._heap:
            raise IndexError("pop from empty heap")
        self._swap(0, len(self._heap) - 1)
        val = self._heap.pop()
        if self._heap:
            self._sift_down(0)
        return val

    def peek(self) -> int:
        return self._heap[0]

    def __len__(self) -> int:
        return len(self._heap)

    def _sift_up(self, idx: int) -> None:
        while idx > 0:
            parent = (idx - 1) // 2
            if self._heap[idx] >= self._heap[parent]:
                break
            self._swap(idx, parent)
            idx = parent

    def _sift_down(self, idx: int) -> None:
        n = len(self._heap)
        while True:
            smallest = idx
            left = 2 * idx + 1
            right = 2 * idx + 2
            if left < n and self._heap[left] < self._heap[smallest]:
                smallest = left
            if right < n and self._heap[right] < self._heap[smallest]:
                smallest = right
            if smallest == idx:
                break
            self._swap(idx, smallest)
            idx = smallest

    def _swap(self, i: int, j: int) -> None:
        self._heap[i], self._heap[j] = self._heap[j], self._heap[i]


def find_kth_largest(nums: List[int], k: int) -> int:
    """Find K-th largest element using a min-heap of size k. O(n log k)."""
    import heapq

    heap: List[int] = []
    for num in nums:
        heapq.heappush(heap, num)
        if len(heap) > k:
            heapq.heappop(heap)
    return heap[0]


def merge_k_sorted_lists(lists: List[List[int]]) -> List[int]:
    """Merge K sorted lists using a priority queue. O(N log k) time."""
    import heapq

    heap: List[tuple[int, int, int]] = []
    for i, lst in enumerate(lists):
        if lst:
            heapq.heappush(heap, (lst[0], i, 0))

    result: List[int] = []
    while heap:
        val, list_idx, elem_idx = heapq.heappop(heap)
        result.append(val)
        if elem_idx + 1 < len(lists[list_idx]):
            heapq.heappush(heap, (lists[list_idx][elem_idx + 1], list_idx, elem_idx + 1))
    return result
